packages/client/wallets/smart-wallet/src/blockchain/wallets/clientDecorator.ts

```diff
@@ -0,0 +1,127 @@
+import { SmartWalletSDKError } from "@/error";
+import { ErrorProcessor } from "@/error/processor";
+import { logInfo } from "@/services/logging";
+import { usesGelatoBundler } from "@/utils/blockchain";
+import { logPerformance } from "@/utils/log";
+import type { SmartAccountClient } from "permissionless";
+import type { EntryPoint } from "permissionless/types/entrypoint";
+import { stringify } from "viem";
+
+import { EVMBlockchainIncludingTestnet } from "@crossmint/common-sdk-base";
+
+const transactionMethods = [
+    "sendTransaction",
+    "writeContract",
+    "sendUserOperation",
+] as const satisfies readonly (keyof SmartAccountClient<EntryPoint>)[];
+
+const signingMethods = [
+    "signMessage",
+    "signTypedData",
+] as const satisfies readonly (keyof SmartAccountClient<EntryPoint>)[];
+
+type TxnMethod = (typeof transactionMethods)[number];
+type SignMethod = (typeof signingMethods)[number];
+
+function isTxnMethod(method: string): method is TxnMethod {
+    return transactionMethods.includes(method as any);
+}
+
+function isSignMethod(method: string): method is SignMethod {
+    return signingMethods.includes(method as any);
+}
+
+/**
+ * A decorator class for SmartAccountClient instances. It enhances the client with:
+ * - Error handling & logging.
+ * - Performance metrics.
+ * - Automatic formatting of transactions for Gelato bundler compatibility.
+ *  */
+export class ClientDecorator {
+    constructor(private readonly errorProcessor: ErrorProcessor) {}
+
+    public decorate<Client extends SmartAccountClient<EntryPoint>>({
+        crossmintChain,
+        smartAccountClient,
+    }: {
+        crossmintChain: EVMBlockchainIncludingTestnet;
+        smartAccountClient: Client;
+    }): Client {
+        return new Proxy(smartAccountClient, {
+            get: (target, prop, receiver) => {
+                const originalMethod = Reflect.get(target, prop, receiver);
+
+                if (
+                    typeof originalMethod !== "function" ||
+                    typeof prop !== "string" ||
+                    !(isSignMethod(prop) || isTxnMethod(prop))
+                ) {
+                    return originalMethod;
+                }
+
+                return (...args: any[]) =>
+                    logPerformance(`CrossmintSmartWallet.${prop}`, () =>
+                        this.execute(target, prop, originalMethod, args, crossmintChain)
+                    );
+            },
+        }) as Client;
+    }
+
+    private async execute<M extends TxnMethod | SignMethod>(
+        target: SmartAccountClient<EntryPoint>,
+        prop: M,
+        // eslint-disable-next-line @typescript-eslint/ban-types
+        originalMethod: Function,
```

Comment:
linter doesn't like this, but this is more descriptive than `any`.

Using `SmartAccountClient<EntryPoint>[M]` doesn't play nicely the proxy code.

Reply:
There may be a better type to use but probably not worth messing with TS for a private method